spec.md

```diff
@@ -0,0 +1,269 @@
+```
+
+Map values can be accessed with the `_[_]` operator, and set with the `_[_<-_]` operator.
+For example, if `m` is the following `Map{Int256,String}`: `3 |-> "hello"   5 |-> "goodbye"`, then:
+
+-   `m[3]` is `"hello"`,
+-   `m[6]` is undefined,
+-   `m[5 <- "hello again"]` is `3 |-> "hello"   5 |-> "hello again"`, and
+-   `m[6 <- "hello again"]` is `3 |-> "hello"   5 |-> "hello"   6 |-> "hello again"`.
+
+### Transitions
+
+State transitions are defined using a similar style to the wasm specification.
+The small-step relation `_=>_` is used, and the execution state is pattern-matched over.
+Uninteresting details of the execution state are abstracted with ellipsis (`...`).
+
+For example, the following defines a simple language for adjusting a `Person`'s age:
+
+```
+    PersonCommand ::= "setAge" Int
+                    | "birthday"
+                    | "switchSides"
+
+    Person ::= Person "<<" PersonCommand
+
+    { ... age : N ... } << setTo N' => { ... age : N'    ... }
+    { ... age : N ... } << birthday => { ... age : N + 1 ... }
+
+    { ... side : Empire     ... } << switchSides => { ... side : Resistance ... }
+    { ... side : Resistance ... } << switchSides => { ... side : Empire     ... }
+```
+
+In this way, the following reductions could be a mid-life crisis:
+
+```
+    { name : "Bob" , age : 35 , side : Empire     } << birthday << switchSides
+    { name : "Bob" , age : 36 , side : Empire     } << switchSides
+    { name : "Bob" , age : 36 , side : Resistance }
+```
+
+Extra requirements can be added as sideconditions using the conditional transition `_=>_if_`:
+
+```
+    PersonCommand ::= "switchSidesIfOlder" Int
+
+    { ... age : N ... } << switchSidesIfOlder N' => { ... age : N ... } << switchSides if N >= N'
+    { ... age : N ... } << switchSidesIfOlder N' => { ... age : N ... }                if N <  N'
+```
+
+Execution State
+---------------
+
+### Overall
+
+The ewasm VM contains state relevant to the EEI functions, as well as a wasm engine.
+
+```
+    EwasmVM ::= { eei    : EEI
+                , engine : WasmVM
+                }
+```
+
+### The wasm VM
+
+The wasm VM is defined in the [wasm spec](https://github.com/webassembly/spec).
+Here we define the fields we need from it.
+
+```
+    WasmVM ::= { stack : InstructionStack }
+```
+
+An `InstructionStack` consists of a space-separate list of `Instruction`.
+
+```
+    InstructionStack ::= ".Stack"
+                       | InstructionStack Instruction
+```
+
+### The EEI
+
+The EEI contains fields relevant to VMs used for executing contracts on the Ethereum blockchain.
+
+**TODO**: In the definition of `EEI.useGas`, the field `gas` is treated like an `i64.const`, but here it's an `Int256`.
+          How do we reconcile these?
+          In EVM, it's an `Int256` in general.
+
+```
+    EEI ::= { statusCode : StatusCode
+            , accounts   : Accounts
+            , gas        : Int256
```

Comment:
I guess this isn't an exhaustive definition of `EEI`?

Reply:
No, I'm just defining the parts of the `EEI` that I need as I add more system contract calls. Only needed these bits for `EEI.useGas`.